File: overRide/2.rb
class Cal
    @@_history = []
    def initialize(v1, v2)
       @v1 = v1
       @v2 = v2
    end

    def Cal.history()
        puts "Cal History"
        for item in @@_history
            p item
        end
    end

    def add()
        result = @v1 + @v2
        @@_history.push("#{@v1} + #{@v2} = #{result}")
        return result
    end
    def subtract()
        result = @v1 - @v2
        @@_history.push("#{@v1} - #{@v2} = #{result}")
        return result
    end
    def info()
        return "Cal => v1: #{@v1}, v2: #{@v2}"
    end
end

class CalMultiply < Cal
    def multiply()
        result = @v1 * @v2
        @@_history.push("#{@v1} * #{@v2} = #{result}")
        return result
    end
    def info()
        return "CalMultiply => #{super()}"
    end
end

class CalDevide < CalMultiply
    def devide()
        result = @v1 / @v2
        @@_history.push("#{@v1} / #{@v2} = #{result}")
        return result
    end
    def info()
        return "CalDevide => #{super()}"
    end
end

c0 = Cal.new(30, 20)
p c0.info()

c1 = CalMultiply.new(50, 30)
p c1.info()

c2 = CalDevide.new(60, 10)
p c2.info()

File: overRide/2.py
class Cal(object):
    _history = []
    def __init__(self, v1, v2):
        if isinstance(v1, int):
            self.v1 = v1
        if isinstance(v2, int):
            self.v2 = v2
    
    def add(self):
        result = self.v1 + self.v2
        Cal._history.append("add: %d + %d = %d" % (self.v1, self.v2, result))
        return result

    def subtract(self):
        result = self.v1 - self.v2
        Cal._history.append("subtract: {} - {} = {}".format(self.v1, self.v2, result))
        return result
    
    @classmethod
    def history(cls):
        for item in cls._history:
            print(item)
    
    def info(self):
        return "Cal => v1 : {}, v2: {}".format(self.v1, self.v2)

class CalMultiply(Cal):
    def multiply(self):
        result = self.v1 * self.v2
        Cal._history.append("multiply: {} * {} = {}".format(self.v1, self.v2, result))
        return result
    def info(self):
        return "CalMultiply => %s" % super().info()

class CalDivide(CalMultiply):
    def devide(self):
        result = self.v1 / self.v2
        Cal._history.append("devide: {} / {} = {}".format(self.v1, self.v2, result))
        return result
    def info(self):
        return "CalDivide => %s" % super().info()

c0 = Cal(30, 60)
print(c0.info())

c1 = CalMultiply(10, 10)
print(c1.info())

c2 = CalDivide(50, 10)
print(c2.info())


